lib/chef/provisioning/aws_driver/managed_aws.rb

```diff
@@ -0,0 +1,231 @@
+        #
+        def lookup_aws_id(type, id, required: false)
+          return nil if id.nil?
+
+          id = id.name if id.is_a?(Chef::Resource)
+
+          case type
+          when :eip_address
+            begin
+              IPAddr.new(id)
+            rescue IPAddr::InvalidAddressError
+              id = get_managed_id(:aws_eip_address, id, required: required)
+            end
+          when :image
+            unless id =~ /^ami-[A-Fa-f0-9]{8}$/
+              image = driver.ec2.images.filter('name', id).first
+              if image && image.exists?
+                id = image.id
+              else
+                raise "No image named #{id} found"
+              end
+            end
+          when :instance
+            unless id =~ /^i-[A-Fa-f0-9]{8}$/
+              id = get_managed_id(:machine, id, key: 'instance_id', required: required)
+            end
+          when :security_group
+            unless id =~ /^sg-[A-Fa-f0-9]{8}$/
+              id = get_managed_id(:aws_security_group, id, required: required)
+            end
+          when :subnet
+            unless id =~ /^subnet-[A-Fa-f0-9]{8}$/
+              id = get_managed_id(:aws_subnet, id, required: required)
+            end
+          when :volume
+            unless id =~ /^vol-[A-Fa-f0-9]{8}$/
+              id = get_managed_id(:aws_volume, id, required: required)
+            end
+          when :vpc
+            unless id =~ /^vpc-[A-Fa-f0-9]{8}$/
+              id = get_managed_id(:aws_vpc, id, required: required)
+            end
+          end
+          id
+        end
+
+        #
+        # Look up the AWS object ID.  Fail if it does not exist.
+        #
+        # @param type The type of AWS object to get.
+        # @param id The ID of the object.
+        #
+        # @return The AWS object id.  If the ID cannot be found, `nil` is returned.
+        #         The method may return ids even if the actual AWS object does not
+        #         exist (it won't look in AWS in all cases).
+        #
+        def lookup_aws_id!(type, id)
+          lookup_aws_id(type, id, required: true)
+        end
+
+        #
+        # Get an AWS object.
+        #
+        # @param type The type of AWS object to get
+        # @param id The ID of the object.
+        # @param required `true` if an error should be raised when the object does not
+        #        exist.  The deepest error possible (such as the 404 response) will be
+        #        raised.  If the input value is `nil`, `nil` will be returned rather
+        #        than an error raised.
+        #
+        # @return The actual AWS object.  If the AWS object doesn't exist, the method
+        #         may either return `nil` or an AWS object where `.exists?` is `false`.
+        #
+        def get_aws_object(type, id, required: false)
+          id = lookup_aws_id(type, id, required: required)
+          if id
+            aws_object = case type
```

Comment:
Again, this case statement feels weird - why can't the individual resources know how to look themselves up?